test/jit/test_union.py

```diff
@@ -0,0 +1,345 @@
+            scripted()
+
+    def test_union_does_not_replace_existing_union_annotated_type(self):
+        def fn():
+            x: List[Union[int, str]] = [1, "foo", 3]
+            x.append(2.0)
+            return x
+
+        with self.assertRaisesRegex(RuntimeError, "Could not match type float"):
+            scripted = torch.jit.script(fn)
+            scripted()
+
+    def test_union_does_not_replace_existing_annotated_type_with_empty_container(self):
+        def fn():
+            x: List[int] = []
+            x.append("foo")
+            return x
+
+        with self.assertRaisesRegex(RuntimeError, "Could not match type str"):
+            scripted = torch.jit.script(fn)
+            scripted()
+
+    def test_unions_of_unions_are_flattened(self):
+        @torch.jit.script
+        def fn_with_union(x: Union[int, str, float]) -> str:
+            return "foo"
+
+        @torch.jit.script
+        def fn_with_nested_union(x: Union[Union[int, str], float]) -> str:
+            return "foo"
+
+        self.assertEqual(fn_with_union(1), fn_with_nested_union(1))
+
+    def test_unions_of_a_single_argument_vanish(self):
+        @torch.jit.script
+        def fn_with_int(x: int) -> str:
+            return "foo"
+
+        @torch.jit.script
+        def fn_with_union_of_int(x: Union[int]) -> str:
+            return "foo"
+
+        self.assertEqual(fn_with_int(1), fn_with_union_of_int(1))
+
+    def test_union_redundant_arguments_are_skipped(self):
+        @torch.jit.script
+        def fn(x: Union[int, str]) -> str:
+            return "foo"
+
+        @torch.jit.script
+        def redundant_fn(x: Union[int, str, int]) -> str:
+            return "foo"
+
+        self.assertEqual(fn(1), redundant_fn(1))
+
+    def test_union_argument_order_is_ignored(self):
+        @torch.jit.script
+        def fn1(x: Union[int, str]) -> str:
+            return "foo"
+
+        @torch.jit.script
+        def fn2(x: Union[str, int]) -> str:
+            return "foo"
+
+        self.assertEqual(fn1(1), fn2(1))
+
+    def test_union_T_None_is_equivalent_to_optional_T(self):
+        def fn(x: Union[int, None]) -> str:
+            return "foo"
+
+        self.checkScript(fn, (1,))
+
+    def test_union_subclasses_larger_union(self):
+        def fn() -> Union[int, str, torch.Tensor]:
+            x: Union[int, str] = "foo"
+            return x
+
+        self.checkScript(fn, ())
+
+    def test_union_with_type_coalescing(self):
+        def fn(x: Union[List[int], int]) -> int:
+            lst = [1, 2, 3]
+            if isinstance(x, int):
+                x = lst
+            return lst[0]
+
+        self.checkScript(fn, (1,))
+        self.checkScript(fn, ([4, 5, 6],))
+
+    def test_union_as_dict_key(self):
+        def fn():
+            x: Dict[Union[int, str], str] = {}
+            x["foo"] = "bar"
+            x[1] = 2
+            return x[1]
+
+        with self.assertRaisesRegex(RuntimeError, "only int, float, "
```

Comment:
Is this a fundamental limitation?

Reply:
No, I want to make this a follow-up issue. I found it harder to do than I expected

Reply:
I see, it makes sense to leave it for later. Could you file a github issue to track it? I don't think this issue is particularly urgent at all, just wanted to ensure it isn't lost.

Also, it would help if you could add `TODO` in appropriate part of code explaining that this is a temporary limitation.

Reply:
Issue [here](https://github.com/pytorch/pytorch/issues/58167)